### PR TITLE
Implement chat messages, mtproto typings and so on

### PR DESCRIPTION
**Pull request by zerobias**
* Implemented mesage slice loading.
Use method api/chats loadSliceRange. It can load various slices ov history, helpful for continious history loading.

* Implemented redux fields chats (possibly should named as dialogs) and users
    *  *chats*. Main dialogs cache, reacts to slice loading, store every dialog messages in normalized form
    *  *users*. Draft realisation. Automatically updates if loaded slice contains unknown users.

* Implemented chat message component. "Dumb" message component also created in Chat folder, because it seems to me that in the future Chat component will become container, and its was better to debug when all child components was located in the same folder.

* Added Mtproto typings. Its really necessary when anybody will add typings to api actions, because almost all api actions begins with server response invoking.

* Moved complex thunk actions out from redux/modules. This dir contains sub-reducers composed to redux state. Possibly it should named
as redux/state or redux/reducers with current reducers.ts as index.ts file.

* Css styling helper classes in helpers/classes

* DC_SERVER env trim

### CODE REVIEW

File: src/app/helpers/Telegram/config.ts [hunk 1]
```diff
@@ -1,9 +1,10 @@
+import { trim } from 'lodash';
```
Comment (goodmind) [inline src/app/helpers/Telegram/config.ts]: Change path to `lodash/trim`

File: src/app/helpers/Telegram/index.ts [hunk 2]
```diff
@@ -95,8 +97,8 @@ export function generateDialogIndex(date) {
 export function getDataCenters() {
   const tl = TypeLanguage;
   const promises = Promise.all([
-    invoke('help.getConfig'),
-    invoke('help.getNearestDc'),
+    invoke<IMtpHelpGetConfig>('help.getConfig'),
```
Comment (goodmind) [inline src/app/helpers/Telegram/index.ts:75]: What about something like this? https://goo.gl/JmBEwO

File: src/app/redux/api/auth.ts [hunk 3]
```diff
@@ -0,0 +1,86 @@
+import { invoke, APP_HASH, APP_ID, makePasswordHash } from 'helpers/Telegram';
```
Comment (goodmind) [inline src/app/redux/api/auth.ts:1]: Why this in `api/auth` folder? Why just don't leave it in `modules/auth` but rename to `actions.ts` or something like this?

File: src/app/redux/modules/messages/index.ts [hunk 4]
```diff
@@ -25,14 +27,14 @@ export const messagesReducer = combineReducers({
 });
 
 export function fetchMessages(peer) {
-  return dispatch => {
+  return (dispatch: IDispatch) => {
     dispatch(MESSAGES.INIT(peer));
-    invoke('messages.getHistory', {
+    invoke<IMtpMessagesSlice>('messages.getHistory', {
```
Comment (goodmind) [inline src/app/redux/modules/messages/index.ts]: Again https://goo.gl/JmBEwO

File: src/app/redux/modules/users.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    ? [ ...ids, ...newList ]
+    : ids;
+};
+
+const ids = createReducer({
+  [LOAD_SLICE.DONE]: (state: number[], { users }: { users: IMtpUser[] }) =>
+    addNewIds(state, users),
+}, []);
+
+// const byId =
```
Comment (goodmind) [inline src/app/redux/modules/users.ts]: Where is users?

**Review by goodmind — CHANGES_REQUESTED** [review]

File: src/app/redux/api/auth.ts (hunk 3 above)
Comment (zerobias) [reply src/app/redux/api/auth.ts:1]: api folder is going to be huge, really bigger than modules folder itself. I don`t know where is should be, but now modules folder seems to be pure store implemenation directory

File: src/app/components/Message/index.tsx [hunk 6]
```diff
@@ -2,7 +2,7 @@ import * as React from 'react';
```
Comment (goodmind) [inline src/app/components/Message/index.tsx]: Why not in `Components/Chat/Message`

File: src/app/helpers/state.ts [hunk 7]
```diff
@@ -25,6 +41,30 @@ export const appendNew = whenNot(contains, append);
...
-export const listToIdMap = P(map(idPair), fromPairs);
+const addChangedProp = <V>(state: TById<V>, [id, val]: [number, V]): TById<V> =>
+  unless(
+    propEq(id, val),
+    assoc(id, val),
+  )(state);
+
+type ITransformer = <F, G>(payload: G) => Array<[number, F]>;
+
+export const changeReducer = <G>(transformer: ITransformer) =>
```
Comment (goodmind) [inline src/app/helpers/state.ts:63]: Weird helper...

File: src/app/redux/modules/dialogs.ts [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+  cleanPayload(dialog),
+);
+
+const addDialog = dialog => R.assoc(getId(dialog), cleanPayload(dialog));
+
+const addChanged = whenNot(isDialogEquals, addDialog);
+
+const byIdOnDialogsDone = getReduce(getDialogs, addChanged);
+
+const byId = createReducer({
```
Comment (goodmind) [inline src/app/redux/modules/dialogs.ts:52]: `createReducer` have type param which represents state

File: src/app/redux/modules/dialogs.ts [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+  [GET_DIALOGS.DONE]: byIdOnDialogsDone,
+}, {});
+
+const reducer = combineReducers({
```
Comment (goodmind) [inline src/app/redux/modules/dialogs.ts:56]: Also generic type param

File: src/app/redux/modules/loadings.ts [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+import { combineReducers } from 'redux';
+
+import { CHATS } from 'redux/actions';
+const { GET_DIALOGS } = CHATS;
+
+export type IStoreLoadings = {
+  chatList: boolean;
+};
+
+const setTrue = () => true;
```
Comment (goodmind) [inline src/app/redux/modules/loadings.ts]: Inconsistent with `modules/auth`, also `ramda` has `R.T` and `R.F`

File: src/app/redux/api/histories.ts [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+
+export interface IDialogPayload {
+  messages: TById<IMtpMessage>;
+  users: IMtpUser[];
+}
+
+export const loadSliceRange = (dispatch: IDispatch) =>
+  (id: number, peer: IMtpPeer, offset: number = 0, limit: number = 10) => {
+    const adapter = (slice: IMtpMessagesSlice) => {
+      const pure = rejectDashAndFuncs(slice);
```
Comment (goodmind) [inline src/app/redux/api/histories.ts]: Why this isn't in `invoke` ?

File: src/app/redux/modules/chatList/index.ts [hunk 12]
```diff
@@ -1,161 +0,0 @@
...
-
-  if (toID < 0) {
-    return toID;
-  } else if (message.out || message.flags & 2) {
-    return toID;
-  }
-  return message.from_id;
-};
-
-const saveConversation = (c, m) => dialog => {
```
Comment (goodmind) [inline src/app/redux/modules/chatList/index.ts:69]: miss

File: src/app/redux/api/chatList.ts [hunk 13]
```diff
@@ -1,14 +1,13 @@
 import { CHATS } from 'actions';
```
Comment (goodmind) [inline src/app/redux/api/chatList.ts:1]: Why `api/histories` mixed with `api/dialogs`?

**Review by goodmind — CHANGES_REQUESTED** [review]
not ok

**Comment by goodmind** [comment]
**Roadmap Updates**

- [ ] Mapped MTProto types (https://goo.gl/JmBEwO)
- [x] R.T and R.F

**Comment by zerobias** [comment]
R.T and R.F are already implemented at https://github.com/goodmind/treact/pull/8/commits/d4e73154617f49a1b16974b6a0d2ff4a0b5428c5

**Pull request merged**